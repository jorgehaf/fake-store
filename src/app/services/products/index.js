import client from "../config/client";

let products = {
    getSingleProduct: (idProduct) => {
        return client.get(`/products/${idProduct}`);
    },
    getProducts: (params) => {
        return client.get("/products", params);
    },
    getCategory: (params) => {
        return client.get("/products/category", params);
    }
};

export default products;
