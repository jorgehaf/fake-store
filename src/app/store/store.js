import { configureStore } from '@reduxjs/toolkit';
import cartReducer from './cartStore';

export const store = configureStore({
  reducer: {
    cart: cartReducer,
  },
});

export default store;
