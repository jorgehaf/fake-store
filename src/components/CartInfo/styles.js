import styled from 'styled-components';

export const CartInfo = styled.div`
  display: flex;
  align-items: center;
  padding: 10px;
  border-radius: 4px;
  border: 1px solid #ccc;
`;

export const CartTotal = styled.span`
  font-weight: bold;
  margin-left: 5px;
`;

const styles = {
    CartInfo,
    CartTotal,
};

export default styles;