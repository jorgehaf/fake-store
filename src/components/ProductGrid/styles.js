import styled from 'styled-components';

export const Grid = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(240px, 1fr));
  gap: 20px;
  padding: 20px;
  width: 80%;
  margin-left: auto;
  margin-right: auto;
`;

export const ProductCard = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  border: 1px solid #ddd;
  border-radius: 8px;
  padding: 10px;
  text-align: center;
  transition: box-shadow 0.2s;

  &:hover {
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
  }
`;

export const ProductImage = styled.img`
  width: 100%;
  height: 150px;
  object-fit: contain;
  margin-bottom: 10px;
  cursor: pointer;
`;

export const ProductName = styled.h3`
  font-size: 16px;
  margin: 10px 0;
  -webkit-line-clamp: 2;
  color: #333;
  text-decoration: none;
  cursor: pointer;
`;

export const CardInfoContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 8px;
`;

export const ProductPrice = styled.p`
  color: green;
  font-weight: bold;
  margin: 5px 0;
`;

export const ProductDescription = styled.p`
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  margin: 10px 0;
`;

const AddToCartButton = styled.button`
  background-color: #444;
  color: white;
  border: none;
  padding: 12px 20px;
  font-size: 16px;
  font-weight: bold;
  border-radius: 8px;
  cursor: pointer;
  transition: background-color 0.3s;
  &:hover {
    background-color: #222;
  }
`;

const styles = {
  Grid,
  ProductCard,
  ProductImage,
  ProductName,
  CardInfoContainer,
  ProductPrice,
  ProductDescription,
  AddToCartButton,
};

export default styles;
