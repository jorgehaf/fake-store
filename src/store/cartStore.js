import { createSlice } from '@reduxjs/toolkit';

const cartStore = createSlice({
  name: 'cart',
  initialState: {
    items: [],
  },
  reducers: {
    addToCart: (state, action) => {
      state.items.push(action.payload);
    },
  },
});

export const { addToCart } = cartStore.actions;

export default cartStore.reducer;
