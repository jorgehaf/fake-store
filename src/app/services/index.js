import products from './products'

const services = {
    products
};

export default services;
