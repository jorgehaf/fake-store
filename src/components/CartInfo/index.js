"use client";

import React from 'react';
import { Provider } from 'react-redux';
import { store } from '@/store/store';

import IndexComponent from './indexComponent'

export default function CartInfo() {

    return (
        <Provider store={store}>
            <IndexComponent />
        </Provider>
    )
}