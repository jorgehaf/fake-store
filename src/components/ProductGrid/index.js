"use client";

import React from 'react';
import { useDispatch } from 'react-redux';
import Link from 'next/link';

import styles from './styles';
import { addToCart } from '@/store/cartStore';

export default function ProductGrid({ filteredProducts }) {
    const dispatch = useDispatch();

    const handleAddToCart = (product) => {
        dispatch(addToCart(product));
    };

    return (
        <styles.Grid>
            {Array.isArray(filteredProducts) && filteredProducts.map((product) => (
                <styles.ProductCard key={product.id}>
                    <Link href={`/products/${product.id}`} passHref legacyBehavior>
                        <a>
                            <styles.ProductImage src={product.image} alt={product.title} />
                            <styles.ProductName>{product.title}</styles.ProductName>
                        </a>
                    </Link>
                    <styles.ProductPrice>R$ {product.price.toFixed(2)}</styles.ProductPrice>
                    <styles.ProductDescription>
                        {product.description.length > 100 ? `${product.description.substring(0, 100)}...` : product.description}
                    </styles.ProductDescription>
                    <styles.AddToCartButton onClick={() => handleAddToCart(product)}>
                        Adicionar ao Carrinho
                    </styles.AddToCartButton>
                </styles.ProductCard>
            ))}
        </styles.Grid>
    )
}