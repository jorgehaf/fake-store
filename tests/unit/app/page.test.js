import { render, screen, fireEvent } from '@testing-library/react';
import Page from '@/app/page';
import { Provider } from 'react-redux';
import { store } from '@/store/store';
import services from '@/services';

import CartInfo from '@/components/CartInfo';
import ProductGrid from '@/components/ProductGrid';


jest.mock('../../../src/services', () => ({
    products: {
        getProducts: jest.fn(),
        getCategory: jest.fn(),
    },
}));

describe('Page Component', () => {
    const productsMock = [
        { id: 1, title: 'Produto 1', price: 10, category: 'Categoria A', description: '' },
        { id: 2, title: 'Produto 2', price: 20, category: 'Categoria B', description: '' },
    ];

    const categoriesMock = ['Categoria A', 'Categoria B'];

    beforeEach(() => {
        services.products.getProducts.mockResolvedValue({
            products: productsMock,
        });

        services.products.getCategory.mockResolvedValue({
            categories: categoriesMock,
        });
    });

    it('Should render Pages and other components', async () => {
        render(
            <Provider store={store}>
                <Page />
            </Provider>
        );

        expect(Page).toMatchSnapshot();
        expect(CartInfo).toMatchSnapshot();
        expect(ProductGrid).toMatchSnapshot();
    });

    it('Should change the product list based on the selected category', async () => {
        render(
            <Provider store={store}>
                <Page />
            </Provider>
        );

        const categoryFilter = await screen.findByRole('combobox');
        fireEvent.change(categoryFilter, { target: { value: 'Categoria A' } });


        expect(services.products.getCategory).toHaveBeenCalledWith({ type: 'Categoria A' });

        expect(ProductGrid).toMatchSnapshot();
    });
});
