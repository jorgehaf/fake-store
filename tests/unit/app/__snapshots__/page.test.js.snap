// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Component Should render Pages and other components 1`] = `[Function]`;

exports[`Page Component Should render Pages and other components 2`] = `[Function]`;

exports[`Page Component Should render Pages and other components 3`] = `[Function]`;

exports[`Page Component deve alterar a lista de produtos com base na categoria selecionada 1`] = `[Function]`;
